feat: Add `num_samples` parameter to `date_range` and `datetime_range`

mcrumiller opened the pull request:
Closes #7525
Closes #23424

This update adds the `num_samples` parameter to `date_range` and `datetime_range`. The input parameters to are now:

* `start`
* `end`
* `interval`
* `num_samples`

A range can be generated using any three of the above:

```python
import polars as pl
from datetime import date

df = pl.DataFrame({
    "start": [date(2025, 1, 1), date(2025, 1, 2)],
    "end": [date(2025, 1, 10), date(2025, 1, 11)],
    "num_samples": [5, 6],
})

df.select(
    start_end_samples=pl.datetime_range(start=col("start").min(), end=col("end").max(), num_samples=6),
    start_interval_samples=pl.datetime_range(start=col("start").min(), interval="-2d", num_samples=col("num_samples").min() + 1),
    end_interval_samples=pl.datetime_range(end=col("end").max(), interval="1mo", num_samples=col("num_samples").max()),
    start_end_interval=pl.datetime_range(start=col("start").min(), end=col("end").max(), interval="2d"),
)
# shape: (6, 4)
# ┌─────────────────────┬────────────────────────┬──────────────────────┬─────────────────────┐
# │ start_end_samples   ┆ start_interval_samples ┆ end_interval_samples ┆ start_end_interval  │
# │ ---                 ┆ ---                    ┆ ---                  ┆ ---                 │
# │ datetime[μs]        ┆ datetime[μs]           ┆ datetime[μs]         ┆ datetime[μs]        │
# ╞═════════════════════╪════════════════════════╪══════════════════════╪═════════════════════╡
# │ 2025-01-01 00:00:00 ┆ 2025-01-01 00:00:00    ┆ 2024-08-11 00:00:00  ┆ 2025-01-01 00:00:00 │
# │ 2025-01-03 00:00:00 ┆ 2024-12-30 00:00:00    ┆ 2024-09-11 00:00:00  ┆ 2025-01-03 00:00:00 │
# │ 2025-01-05 00:00:00 ┆ 2024-12-28 00:00:00    ┆ 2024-10-11 00:00:00  ┆ 2025-01-05 00:00:00 │
# │ 2025-01-07 00:00:00 ┆ 2024-12-26 00:00:00    ┆ 2024-11-11 00:00:00  ┆ 2025-01-07 00:00:00 │
# │ 2025-01-09 00:00:00 ┆ 2024-12-24 00:00:00    ┆ 2024-12-11 00:00:00  ┆ 2025-01-09 00:00:00 │
# │ 2025-01-11 00:00:00 ┆ 2024-12-22 00:00:00    ┆ 2025-01-11 00:00:00  ┆ 2025-01-11 00:00:00 │
# └─────────────────────┴────────────────────────┴──────────────────────┴─────────────────────┘

Still a WIP, but getting there.
```

mcrumiller commented:
@MarcoGorelli it's almost there, but I'm a bit confused of what intended behavior should be. `int_range` doesn't have a `closed` parameter. I think that for continuous quantities like datetime (I know it's technically discrete, but it represents a continuous quantity), having `closed` makes sense, but for discrete units like datetime, I am not so sure.

Suppose we had the closed interval for `int_range`. What does `int_range(1, 10, step=3, closed="right")` look like? Does it go 2, 5, 8, as in it starts as close to 1 as possible, but not exactly on 1 since it's right-closed? Or does it go 4, 7, 10, i.e. the left-most value is left out? The same holds for dates here. `date_range(start="2025-01-01", end="2025-01-10", num_samples=3, closed="right")`. What's the first value?

etiennebacher commented:
> This update adds the `num_samples` (I opted for this name over `periods` to be more in line with our range functions) parameter to `date_range` and `datetime_range`

I find this argument name very unclear about what it does. If you're looking for something that can be applied to all range functions, maybe `length_output` would do the job? `pl.date_range(start=date(1985, 1, 1), end=date(1985, 1, 10), length_output=5)` looks explicit enough to me.

(I'm not sure which linked issue is more appropriate to discuss the API design, which is why I comment here)

mcrumiller commented:
@MarcoGorelli almost there, question question. Right now, `datetime_range(date, date)` returns datetime with `us` time units. Do we want to use `ms` instead, since that makes it the least likely that we'll out of bounds on conversion? My implementation converts dates to `ms` but I can switch to `us` if we want to stay consistent.

mcrumiller commented:
@MarcoGorelli I think it's finally good to go. Any suggestions are welcome. It ended up being a bit more of an undertaking than I had initially anticipated, so I did my best maximize test coverage (thank you Coverage Test).

There are two main changes that affect the code outside of just adding the parameters:

1. All casts and time zone conversions are moved to the IR stage, which I think is the preferred way to do things around here.
2. The default output unit in `datetime_range` and `datetime_ranges` when the input is a `pl.Date` is now milliseconds instead of microseconds.

Aside from that, I think usage is the same. Let me know if you have any concerns. Thanks!

mcrumiller commented on py-polars/tests/unit/operations/rolling/test_rolling.py:
```diff
@@ -1624,9 +1626,11 @@ def test_rolling_sum_non_finite_23115(with_nulls: bool) -> None:
     data = random.choices(values, k=1000)
     naive = [
-        sum(0 if x is None else x for x in data[max(0, i + 1 - 4) : i + 1])
-        if sum(x is not None for x in data[max(0, i + 1 - 4) : i + 1]) >= 2
-        else None
+        (
+            sum(0 if x is None else x for x in data[max(0, i + 1 - 4) : i + 1])
+            if sum(x is not None for x in data[max(0, i + 1 - 4) : i + 1]) >= 2
+            else None
+        )
```
I am not sure how this one got in here, sorry. I don't see any other non-functional changes though, am I missing some?

MarcoGorelli replied:
ah no worries, looks like this was the only one, the others are legit diffs sorry - thanks!

MarcoGorelli commented on py-polars/polars/functions/range/datetime_range.py:
```diff
@@ -21,22 +21,24 @@
     closed: ClosedInterval = ...,
     time_unit: TimeUnit | None = ...,
     time_zone: str | None = ...,
-    eager: Literal[False] = ...,
+    eager: Literal[False],
```
quick comment `Literal[False]` is the default, so `= ...` looks correct, no need to remove it

MarcoGorelli replied:
`= ...` means that the given annotation includes the default value. So this is both for when the user passes `eager=False` and when they don't specify `eager` at all

mcrumiller replied:
Fixed in [3a16924](https://github.com/pola-rs/polars/pull/23425/commits/3a16924a30004a389a4eae23ea7b520538c3f369).

mcrumiller commented:
@MarcoGorelli this one should be ready again, after incorporating the other PR.

MarcoGorelli commented on py-polars/polars/functions/range/datetime_range.py:
```diff
@@ -47,21 +49,23 @@ def datetime_range(
+    num_samples: int | IntoExprColumn | None = None,
     *,
     closed: ClosedInterval = ...,
     time_unit: TimeUnit | None = ...,
     time_zone: str | None = ...,
-    eager: bool,
+    eager: bool = ...,
 ) -> Series | Expr: ...
```
I don't think this is accurate. If the default for `eager` is passed (`False`), then we know the output type will be `Expr`. I think this overload was probably specifically for the case when `eager` is a variable of type `bool`, and so the return type could be either `Series` or `Expr`

mcrumiller replied:
Yep, my bad, this was from before I knew what the `...` meant.

mcrumiller replied:
[937a3ae](https://github.com/pola-rs/polars/pull/23425/commits/937a3ae9d7775779bdb7afc4c66d1b5644722423)

MarcoGorelli commented on py-polars/polars/functions/range/datetime_range.py:
```diff
@@ -254,21 +307,23 @@ def datetime_ranges(
 
 @overload
 def datetime_ranges(
-    start: datetime | date | IntoExprColumn,
-    end: datetime | date | IntoExprColumn,
-    interval: str | timedelta = ...,
+    start: datetime | date | IntoExprColumn | None = None,
+    end: datetime | date | IntoExprColumn | None = None,
+    interval: str | timedelta | None = None,
+    num_samples: int | IntoExprColumn | None = None,
     *,
     closed: ClosedInterval = ...,
     time_unit: TimeUnit | None = ...,
     time_zone: str | None = ...,
-    eager: bool,
+    eager: bool = ...,
```
same

mcrumiller replied:
[937a3ae](https://github.com/pola-rs/polars/pull/23425/commits/937a3ae9d7775779bdb7afc4c66d1b5644722423)

MarcoGorelli commented on py-polars/polars/functions/range/date_range.py:
```diff
@@ -291,11 +364,27 @@ def date_ranges(
+        and num_samples is not None
+        and closed != "both"
+    ):
+        msg = (
+            "date_ranges does not support 'left', 'right', or 'none' for the 'closed' "
+            "parameter when 'start', 'end', and 'num_samples' is provided."
```
same comment as earlier on the message

MarcoGorelli commented on py-polars/polars/functions/range/date_range.py:
```diff
@@ -207,26 +269,31 @@ def date_ranges(
     """
     Create a column of date ranges.
 
+    Exactly three of 'start', 'end', 'interval', and 'periods' must be provided to
+    construct the column of date ranges.
```
same

mcrumiller replied:
[937a3ae](https://github.com/pola-rs/polars/pull/23425/commits/937a3ae9d7775779bdb7afc4c66d1b5644722423)

MarcoGorelli commented on crates/polars-time/src/upsample.rs:
```diff
@@ -122,59 +122,65 @@ fn upsample_impl(
 ) -> PolarsResult<DataFrame> {
     let s = source.column(index_column)?;
     let time_type = s.dtype();
-    if matches!(time_type, DataType::Date) {
```
the diff in this file is a little hard to parse, could you summarise the gist of the change please?

i'm not sure I see why `upsample_impl` needs to change

MarcoGorelli commented on crates/polars-plan/src/plans/conversion/type_coercion/datetime_range.rs:
```diff
@@ -0,0 +1,215 @@
+    };
+}
+
+macro_rules! extract_date {
+    ($input:expr, $expr_arena:expr, $schema:expr, $idx:literal, $arg:literal) => {{
+        let (_, dtype) =
+            unpack!(get_aexpr_and_type($expr_arena, $input[$idx].node(), $schema));
+        polars_ensure!(
+            matches!(dtype, DataType::Datetime(_, _) | DataType::Date),
+            ComputeError: "'{}' must be Date or Datetime, got {:?}", $arg, dtype
+        );
+        dtype
+    }}
+}
+
+macro_rules! extract_samples {
+    ($input:expr, $expr_arena:expr, $schema:expr, $idx:literal) => {{
+        let (_, dtype) =
+        unpack!(get_aexpr_and_type($expr_arena, $input[$idx].node(), $schema));
+        polars_ensure!(
+            dtype.is_integer(),
+            ComputeError: "'num_samples' must be integer, got {:?}", dtype
+        );
+        dtype
+    }}
+}
```
do these have to be macros, or could they be functions?

mcrumiller commented on crates/polars-plan/src/dsl/function_expr/range.rs:
```diff
@@ -7,7 +7,47 @@ use polars_time::{ClosedWindow, Duration};
+            (None, Some(end), Some(_), Some(num_samples)) => {
+                Ok((vec![end, num_samples], DateRangeArgs::EndIntervalSamples))
+            },
+            _ => {
+                polars_bail!(InvalidOperation: "Exactly three of 'start', 'end', 'interval', and 'num_samples' must be supplied.");
+            },
```
We do actually need three parameters defined to create a range. As a special case for backwards compatibility, we default to `interval="1d"` when only `start`/`end` are used.

I can change the rules here, and we can default to 1 day in all cases where only two inputs are supplied. Alternatively, I could update the message to say three are required, except for when only `start`/`end` are used, in which case we default to 1 day.

What do you think?

mcrumiller commented on crates/polars-plan/src/dsl/functions/range.rs:
```diff
@@ -32,82 +32,153 @@ pub fn int_ranges(start: Expr, end: Expr, step: Expr, dtype: impl Into<DataTypeE
+    // Only "both" is supported for date_range(start, end, num_samples).
+    polars_ensure!(
+        !(arg_type == DateRangeArgs::StartEndSamples && closed != ClosedWindow::Both),
+        InvalidOperation: "date_range does not support 'left', 'right', or 'none' for the \
+            'closed' parameter when 'start', 'end', and 'num_samples' is provided.",
```
Well, there are four parameters, as `interval` is one of the parameters.

mcrumiller commented:
> One question I have is whether the changes in `crates/polars-plan/src/plans/conversion/type_coercion/mod.rs` could be factored out as a separate pr?

This could definitely be done, as it's very similar code is used in #23786. Do you want me to go ahead and do that? It may be a bit more work for me since I will then have to re-factor this PR, but I'm willing to do that.

mcrumiller commented on crates/polars-plan/src/dsl/functions/range.rs:
```diff
@@ -32,82 +32,153 @@ pub fn int_ranges(start: Expr, end: Expr, step: Expr, dtype: impl Into<DataTypeE
 }
 
-/// Create a column of datetime ranges from a `start` and `stop` expression.
+/// Create a datetime range from `start`, `stop`, `interval`, and `num_sample` expressions.
```
[a767ddf](https://github.com/pola-rs/polars/pull/23425/commits/a767ddf6eb1fc6875a4c3b176e02a647a0c9184a)

mcrumiller commented on py-polars/polars/functions/range/date_range.py:
```diff
@@ -43,26 +46,31 @@ def date_range(
     Generate a date range.
 
+    Exactly three of 'start', 'end', 'interval', and 'num_samples' mus be provided to
+    construct the date range.
```
[937a3ae](https://github.com/pola-rs/polars/pull/23425/commits/937a3ae9d7775779bdb7afc4c66d1b5644722423)

mcrumiller commented on py-polars/tests/unit/operations/aggregation/test_aggregations.py:
```diff
@@ -191,14 +191,17 @@ def test_duration_function_literal() -> None:
+                    ],
                 ],
-            ],
+                dtype=pl.List(pl.Duration("us")),
```
[94cd993](https://github.com/pola-rs/polars/pull/23425/commits/94cd9937fdcf969f2840a13256e01db963af40f4)

mcrumiller commented on py-polars/tests/unit/operations/rolling/test_rolling.py:
```diff
@@ -852,7 +852,9 @@ def test_rolling_aggregations_with_over_11225() -> None:
-            "date": pl.datetime_range(date(2001, 1, 1), date(2001, 1, 5), eager=True),
+            "date": pl.datetime_range(
+                date(2001, 1, 1), date(2001, 1, 5), time_unit="us", eager=True
+            ),
```
[94cd993](https://github.com/pola-rs/polars/pull/23425/commits/94cd9937fdcf969f2840a13256e01db963af40f4)

mcrumiller commented:
@MarcoGorelli I think I've addressed most of your issue from today. The two issues remaining are:

1. [this](https://github.com/pola-rs/polars/pull/23425#discussion_r2280306426) and similar comments regarding the invalid operation message. I've updated the docstring to now say that while a range requires 3 values to construct, we infer "1d" if only two are provided (as long as one of the two is not the interval, of course). IMO this should be sufficient, but if you think not, we can try to work on improving clarity some more.

2. The IR casts--want me to make a separate PR for that?

nameexhaustion commented:
Could we break the changes down into separate PRs? E.g. -

* PR to add `num_samples` to the DSL / IR. This won't be connected yet to the API and will be ignored by execution.
* PR to add execution-level logic, Python API parameter and test cases

For the above, please perform the changes in-place (i.e. avoid moving code / files around), this will make the diff much easier to review.

The refactoring to move/split code should be done afterwards. Alternatively, you can move the existing code without changing it as the very first step - the goal is to simply ensure that the diff is able to show meaningful differences rather than being caused by code shifting / moving.
